In [1]:
import tensorflow as tf
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import Dense, Flatten
from tensorflow.keras.utils import to_categorical
import matplotlib.pyplot as plt
import numpy as np


# Téléchargement des données MNIST
(x_train, y_train), (x_test, y_test) = tf.keras.datasets.mnist.load_data()

# Normalisation des pixels (les valeurs passent de 0-255 à 0-1)
x_train = x_train / 255.0
x_test = x_test / 255.0

# Aplatissement des images (28x28 -> 784) comme demandé dans le TP
x_train_flat = x_train.reshape(-1, 784)
x_test_flat = x_test.reshape(-1, 784)

# Encodage des labels en one-hot (nécessaire pour la fonction categorical_crossentropy)
y_train_cat = to_categorical(y_train, 10)
y_test_cat = to_categorical(y_test, 10)

print("Format des données d'entraînement :", x_train_flat.shape)



11490434/11490434 ━━━━━━━━━━━━━━━━━━━━ 0s 0us/step
Format des données d'entraînement : (60000, 784)


In [10]:
#1. Choix de la Fonction d’Activation
#création du premier réseau à partir de la fonction d’activation softmax :

# Création d'un modèle de réseau de neurones séquentiel
model_a = Sequential([
    # Couche Dense (entièrement connectée) avec 10 neurones
    # activation='softmax' transforme les sorties en probabilités
    # input_shape=(784,) signifie que chaque image d'entrée est un vecteur de 784 valeurs (28x28 pixels aplatis)
    Dense(10, activation='softmax', input_shape=(784,))
])

# Compilation du modèle : configuration de l'apprentissage
model_a.compile(
    optimizer='adam',                  # Algorithme d'optimisation Adam pour ajuster les poids
    loss='categorical_crossentropy',   # Fonction de perte adaptée à la classification multi-classes
    metrics=['accuracy']               # Mesure de performance : pourcentage de bonnes prédictions
)

# Entraînement du modèle sur les données d'entraînement
history_a = model_a.fit(
    x_train_flat,      # images d'entraînement aplaties en vecteurs de 784 valeurs
    y_train_cat,       # labels convertis en format catégoriel (one-hot encoding)
    epochs=10,         # le modèle voit tout le dataset 10 fois
    batch_size=32,     # 32 exemples sont traités avant mise à jour des poids
    validation_split=0.2  # 20% des données d'entraînement sont utilisées pour la validation
)


Epoch 1/10
1500/1500 ━━━━━━━━━━━━━━━━━━━━ 3s 2ms/step - accuracy: 0.7918 - loss: 0.7968 - val_accuracy: 0.9117 - val_loss: 0.3179
Epoch 2/10
1500/1500 ━━━━━━━━━━━━━━━━━━━━ 3s 2ms/step - accuracy: 0.9105 - loss: 0.3236 - val_accuracy: 0.9207 - val_loss: 0.2876
Epoch 3/10
1500/1500 ━━━━━━━━━━━━━━━━━━━━ 3s 2ms/step - accuracy: 0.9164 - loss: 0.2971 - val_accuracy: 0.9224 - val_loss: 0.2752
Epoch 4/10
1500/1500 ━━━━━━━━━━━━━━━━━━━━ 5s 2ms/step - accuracy: 0.9200 - loss: 0.2832 - val_accuracy: 0.9277 - val_loss: 0.2685
Epoch 5/10
1500/1500 ━━━━━━━━━━━━━━━━━━━━ 2s 2ms/step - accuracy: 0.9237 - loss: 0.2749 - val_accuracy: 0.9262 - val_loss: 0.2678
Epoch 6/10
1500/1500 ━━━━━━━━━━━━━━━━━━━━ 3s 2ms/step - accuracy: 0.9247 - loss: 0.2694 - val_accuracy: 0.9282 - val_loss: 0.2651
Epoch 7/10
1500/1500 ━━━━━━━━━━━━━━━━━━━━ 2s 2ms/step - accuracy: 0.9270 - loss: 0.2594 - val_accuracy: 0.9274 - val_loss: 0.2651
Epoch 8/10
1500/1500 ━━━━━━━━━━━━━━━━━━━━ 3s 2ms/step - accuracy: 0.9283 - loss: 0.2577 - 

In [9]:
# b. Implémentation avec 2 couches cachées et la fonction d’activation ReLU

# Création d’un modèle séquentiel (les couches sont empilées les unes après les autres)
model_b = Sequential([

    # Première couche Dense avec 128 neurones
    # input_shape=(784,) : vecteur d'entrée correspondant à une image 28x28 aplatie
    # activation='relu' : fonction ReLU (max(0,x)) qui introduit de la non-linéarité
    Dense(128, activation='relu', input_shape=(784,)),

    # Deuxième couche cachée avec 64 neurones
    # ReLU est encore utilisée pour améliorer l’apprentissage des relations complexes
    Dense(64, activation='relu'),

    # Couche de sortie avec 10 neurones (une pour chaque classe)
    # Softmax transforme les sorties en probabilités dont la somme vaut 1
    Dense(10, activation='softmax')
])

# Compilation du modèle
model_b.compile(
    optimizer='adam',                 # Optimiseur Adam pour ajuster les poids
    loss='categorical_crossentropy',  # Fonction de perte adaptée à la classification multi-classes
    metrics=['accuracy']              # Mesure de performance : taux de bonnes prédictions
)

# Entraînement du réseau de neurones
history_b = model_b.fit(
    x_train_flat,        # Données d'entraînement (images aplaties en vecteurs de 784)
    y_train_cat,         # Labels encodés en one-hot
    epochs=10,           # Le modèle parcourt le dataset 10 fois
    batch_size=32,       # 32 exemples sont utilisés avant chaque mise à jour des poids
    validation_split=0.2 # 20% des données sont utilisées pour valider le modèle pendant l'entraînement
)

# Affichage du nombre total de paramètres (poids + biais) du réseau
print("Synapses Modèle B:", model_b.count_params())

Epoch 1/10
1500/1500 ━━━━━━━━━━━━━━━━━━━━ 6s 3ms/step - accuracy: 0.8722 - loss: 0.4423 - val_accuracy: 0.9527 - val_loss: 0.1529
Epoch 2/10
1500/1500 ━━━━━━━━━━━━━━━━━━━━ 4s 3ms/step - accuracy: 0.9655 - loss: 0.1154 - val_accuracy: 0.9674 - val_loss: 0.1087
Epoch 3/10
1500/1500 ━━━━━━━━━━━━━━━━━━━━ 5s 3ms/step - accuracy: 0.9766 - loss: 0.0764 - val_accuracy: 0.9706 - val_loss: 0.0988
Epoch 4/10
1500/1500 ━━━━━━━━━━━━━━━━━━━━ 4s 3ms/step - accuracy: 0.9828 - loss: 0.0546 - val_accuracy: 0.9722 - val_loss: 0.0969
Epoch 5/10
1500/1500 ━━━━━━━━━━━━━━━━━━━━ 4s 3ms/step - accuracy: 0.9877 - loss: 0.0392 - val_accuracy: 0.9741 - val_loss: 0.0864
Epoch 6/10
1500/1500 ━━━━━━━━━━━━━━━━━━━━ 5s 4ms/step - accuracy: 0.9895 - loss: 0.0313 - val_accuracy: 0.9749 - val_loss: 0.0915
Epoch 7/10
1500/1500 ━━━━━━━━━━━━━━━━━━━━ 4s 3ms/step - accuracy: 0.9926 - loss: 0.0225 - val_accuracy: 0.9717 - val_loss: 0.1058
Epoch 8/10
1500/1500 ━━━━━━━━━━━━━━━━━━━━ 6s 4ms/step - accuracy: 0.9917 - loss: 0.0245 - 

In [11]:
# c. Implémentation avec 1 couche cachée et la fonction d’activation tanh

# Création d’un modèle séquentiel (les couches sont empilées les unes après les autres)
model_c = Sequential([

    # Première couche Dense (couche cachée) avec 128 neurones
    # input_shape=(784,) : vecteur d’entrée représentant une image 28x28 aplatie (784 pixels)
    # activation='tanh' : fonction d’activation tangente hyperbolique (sorties entre -1 et 1)
    Dense(128, activation='tanh', input_shape=(784,)),

    # Couche de sortie avec 10 neurones (correspondant aux 10 classes possibles)
    # Softmax transforme les valeurs en probabilités pour chaque classe
    Dense(10, activation='softmax')
])

# Compilation du modèle : configuration de l’apprentissage
model_c.compile(
    optimizer='adam',                 # Optimiseur Adam pour ajuster les poids du réseau
    loss='categorical_crossentropy',  # Fonction de perte adaptée à la classification multi-classes
    metrics=['accuracy']              # Mesure de performance : taux de prédictions correctes
)

# Entraînement du modèle sur les données
history_c = model_c.fit(
    x_train_flat,        # Données d'entraînement (images converties en vecteurs de 784 valeurs)
    y_train_cat,         # Labels encodés en format catégoriel (one-hot encoding)
    epochs=10,           # Le modèle parcourt l'ensemble du dataset 10 fois
    batch_size=32,       # Les poids sont mis à jour toutes les 32 observations
    validation_split=0.2 # 20% des données sont utilisées pour valider le modèle pendant l'entraînement
)

# Affichage du nombre total de paramètres (poids + biais) du modèle
print("Synapses Modèle C:", model_c.count_params())

Epoch 1/10
1500/1500 ━━━━━━━━━━━━━━━━━━━━ 5s 3ms/step - accuracy: 0.8641 - loss: 0.4684 - val_accuracy: 0.9478 - val_loss: 0.1841
Epoch 2/10
1500/1500 ━━━━━━━━━━━━━━━━━━━━ 5s 3ms/step - accuracy: 0.9506 - loss: 0.1653 - val_accuracy: 0.9607 - val_loss: 0.1396
Epoch 3/10
1500/1500 ━━━━━━━━━━━━━━━━━━━━ 4s 3ms/step - accuracy: 0.9675 - loss: 0.1114 - val_accuracy: 0.9660 - val_loss: 0.1143
Epoch 4/10
1500/1500 ━━━━━━━━━━━━━━━━━━━━ 4s 3ms/step - accuracy: 0.9772 - loss: 0.0781 - val_accuracy: 0.9710 - val_loss: 0.1007
Epoch 5/10
1500/1500 ━━━━━━━━━━━━━━━━━━━━ 5s 3ms/step - accuracy: 0.9832 - loss: 0.0583 - val_accuracy: 0.9685 - val_loss: 0.1048
Epoch 6/10
1500/1500 ━━━━━━━━━━━━━━━━━━━━ 5s 3ms/step - accuracy: 0.9877 - loss: 0.0461 - val_accuracy: 0.9746 - val_loss: 0.0859
Epoch 7/10
1500/1500 ━━━━━━━━━━━━━━━━━━━━ 5s 4ms/step - accuracy: 0.9918 - loss: 0.0325 - val_accuracy: 0.9741 - val_loss: 0.0880
Epoch 8/10
1500/1500 ━━━━━━━━━━━━━━━━━━━━ 5s 3ms/step - accuracy: 0.9944 - loss: 0.0245 - 

In [12]:
# d. Implémentation avec 1 couche cachée et la fonction d’activation sigmoid

# Création d’un modèle séquentiel (les couches sont empilées les unes après les autres)
model_d = Sequential([

    # Première couche Dense (couche cachée) avec 128 neurones
    # input_shape=(784,) : vecteur d’entrée correspondant à une image 28x28 aplatie (784 pixels)
    # activation='sigmoid' : fonction d’activation sigmoïde qui produit des valeurs entre 0 et 1
    Dense(128, activation='sigmoid', input_shape=(784,)),

    # Couche de sortie avec 10 neurones correspondant aux 10 classes possibles
    # Softmax convertit les sorties en probabilités pour chaque classe
    Dense(10, activation='softmax')
])

# Compilation du modèle : configuration du processus d’apprentissage
model_d.compile(
    optimizer='adam',                 # Optimiseur Adam pour ajuster les poids du réseau
    loss='categorical_crossentropy',  # Fonction de perte adaptée à la classification multi-classes
    metrics=['accuracy']              # Mesure de performance : précision des prédictions
)

# Entraînement du modèle sur les données d’apprentissage
history_d = model_d.fit(
    x_train_flat,        # Images d'entraînement transformées en vecteurs de 784 valeurs
    y_train_cat,         # Labels encodés en format catégoriel (one-hot encoding)
    epochs=10,           # Le modèle parcourt le dataset complet 10 fois
    batch_size=32,       # Les poids sont mis à jour après chaque lot de 32 exemples
    validation_split=0.2 # 20% des données sont utilisées pour la validation
)

# Affichage du nombre total de paramètres (poids + biais) du réseau
print("Synapses Modèle D:", model_d.count_params())

Epoch 1/10
1500/1500 ━━━━━━━━━━━━━━━━━━━━ 5s 3ms/step - accuracy: 0.8203 - loss: 0.7255 - val_accuracy: 0.9257 - val_loss: 0.2521
Epoch 2/10
1500/1500 ━━━━━━━━━━━━━━━━━━━━ 5s 4ms/step - accuracy: 0.9308 - loss: 0.2420 - val_accuracy: 0.9489 - val_loss: 0.1865
Epoch 3/10
1500/1500 ━━━━━━━━━━━━━━━━━━━━ 6s 4ms/step - accuracy: 0.9488 - loss: 0.1764 - val_accuracy: 0.9572 - val_loss: 0.1532
Epoch 4/10
1500/1500 ━━━━━━━━━━━━━━━━━━━━ 5s 3ms/step - accuracy: 0.9622 - loss: 0.1304 - val_accuracy: 0.9596 - val_loss: 0.1388
Epoch 5/10
1500/1500 ━━━━━━━━━━━━━━━━━━━━ 4s 3ms/step - accuracy: 0.9686 - loss: 0.1116 - val_accuracy: 0.9658 - val_loss: 0.1174
Epoch 6/10
1500/1500 ━━━━━━━━━━━━━━━━━━━━ 4s 3ms/step - accuracy: 0.9754 - loss: 0.0878 - val_accuracy: 0.9692 - val_loss: 0.1082
Epoch 7/10
1500/1500 ━━━━━━━━━━━━━━━━━━━━ 5s 3ms/step - accuracy: 0.9804 - loss: 0.0716 - val_accuracy: 0.9707 - val_loss: 0.1009
Epoch 8/10
1500/1500 ━━━━━━━━━━━━━━━━━━━━ 4s 3ms/step - accuracy: 0.9830 - loss: 0.0617 - 

1. Comparaison des performances (Accuracy et Loss)

Modèle A (Softmax, sans couche cachée) : précision d’environ 93 %, la plus faible. L’absence de couches cachées limite la capacité du modèle à apprendre des relations complexes.

Modèle B (ReLU, 2 couches cachées) : meilleure précision (~97.5 %), mais un léger surapprentissage apparaît lorsque la validation loss commence à augmenter.

Modèle C (Tanh, 1 couche cachée) : très bonne précision (~97.4 %) avec la loss la plus faible, montrant un apprentissage stable.

Modèle D (Sigmoid, 1 couche cachée) : bonne précision (~97.3 %) mais une loss légèrement plus élevée que le modèle C.

2. Vitesse de convergence (epochs)

ReLU (B) et Tanh (C) convergent le plus rapidement, dépassant 96 % de précision dès l’epoch 3.

Sigmoid (D) converge plus lentement au début à cause de gradients plus faibles.

Softmax seul (A) stagne rapidement car l’architecture est trop simple.

3. Nombre de synapses (complexité et mémoire)

Modèle A (7 850 synapses) : très léger et rapide, mais moins précis.

Modèles C et D (101 770 synapses) : compromis efficace entre complexité et précision.

Modèle B (109 386 synapses) : le plus complexe, mais le gain de précision est faible par rapport à C et D.


In [5]:
# 2. Choix de l’Algorithme d’Optimisation
# Initialisation d'un dictionnaire pour stocker les performances de chaque modèle
resultats_globaux = {}

# Liste des optimiseurs à évaluer [cite: 40]
liste_optimiseurs = ['sgd', 'adam', 'rmsprop', 'adagrad']

for optimiseur in liste_optimiseurs:
    print(f"\nDébut de l'entraînement avec l'optimiseur : {optimiseur.upper()}...")

    # Création de l'architecture de base (1 couche cachée, 128 neurones, ReLU)
    model = Sequential([
        Dense(128, activation='relu', input_shape=(784,)),
        Dense(10, activation='softmax')
    ])

    # Compilation avec l'optimiseur spécifique à cette itération
    model.compile(optimizer=optimiseur,
                  loss='categorical_crossentropy',
                  metrics=['accuracy'])

    # Lancement de l'apprentissage (verbose=0 masque les barres de progression pour plus de lisibilité)
    history = model.fit(x_train_flat, y_train_cat,
                        epochs=10,
                        batch_size=32,
                        validation_split=0.2,
                        verbose=0)

    # Extraction des résultats de la dernière epoch
    precision_finale = history.history['val_accuracy'][-1]
    perte_finale = history.history['val_loss'][-1]

    # Sauvegarde des données dans le dictionnaire pour le résumé final
    resultats_globaux[optimiseur] = {
        'Précision (Val)': precision_finale,
        'Perte (Val)': perte_finale
    }

    print(f"-> Terminé. Précision: {precision_finale:.4f} | Perte: {perte_finale:.4f}")

# ==========================================
# GÉNÉRATION DU RÉSUMÉ DE TOUTES LES DONNÉES
# ==========================================
print("\n" + "="*60)
print(" RÉSUMÉ COMPARATIF DES OPTIMISEURS (APRÈS 10 EPOCHS)")
print("="*60)

# Affichage formaté des résultats stockés
for optimiseur, métriques in resultats_globaux.items():
    # Ajout des compromis théoriques en commentaire pour justifier les résultats obtenus
    if optimiseur == 'sgd':
        description = "Base simple, classique, intuition sur descente de gradient"
    elif optimiseur == 'adam':
        description = "Moderne, converge vite, combine momentum + RMSprop"
    elif optimiseur == 'rmsprop':
        description = "Adaptation du taux d'apprentissage, stabilité sur entrées variées"
    elif optimiseur == 'adagrad':
        description = "Learning rate par paramètre, illustration de l’effet adaptatif"

    print(f"[{optimiseur.upper():<7}] Précision: {métriques['Précision (Val)']:.4f} | Perte: {métriques['Perte (Val)']:.4f}")
    print(f"          Compromis: {description}\n")

print("="*60)


Début de l'entraînement avec l'optimiseur : SGD...
-> Terminé. Précision: 0.9486 | Perte: 0.1866

Début de l'entraînement avec l'optimiseur : ADAM...
-> Terminé. Précision: 0.9753 | Perte: 0.0990

Début de l'entraînement avec l'optimiseur : RMSPROP...
-> Terminé. Précision: 0.9767 | Perte: 0.0979

Début de l'entraînement avec l'optimiseur : ADAGRAD...
-> Terminé. Précision: 0.9115 | Perte: 0.3339

 RÉSUMÉ COMPARATIF DES OPTIMISEURS (APRÈS 10 EPOCHS)
[SGD    ] Précision: 0.9486 | Perte: 0.1866
          Compromis: Base simple, classique, intuition sur descente de gradient

[ADAM   ] Précision: 0.9753 | Perte: 0.0990
          Compromis: Moderne, converge vite, combine momentum + RMSprop

[RMSPROP] Précision: 0.9767 | Perte: 0.0979
          Compromis: Adaptation du taux d'apprentissage, stabilité sur entrées variées

[ADAGRAD] Précision: 0.9115 | Perte: 0.3339
          Compromis: Learning rate par paramètre, illustration de l’effet adaptatif



In [6]:
# 3. Choix de la Fonction Coût
# Préparation des labels spécifiques pour tester les différentes fonctions de coût
# y_train_cat : Labels one-hot pour categorical_crossentropy [cite: 44]
# y_train : Labels entiers (0-9) pour sparse_categorical_crossentropy [cite: 44]
# y_train_bin : Création de labels binaires (ex: 1 si le chiffre est 5, 0 sinon) pour binary_crossentropy [cite: 44]
y_train_bin = (y_train == 5).astype(int)
y_test_bin = (y_test == 5).astype(int)

resultats_couts = {}

# ==========================================
# 1. binary_crossentropy
# ==========================================
print("Entraînement avec binary_crossentropy...")
model_bin = Sequential([
    Dense(128, activation='relu', input_shape=(784,)),
    # Sortie : 1 neurone avec sigmoid pour une probabilité 0 ≤ y ≤ 1 [cite: 44]
    Dense(1, activation='sigmoid')
])
model_bin.compile(optimizer='adam', loss='binary_crossentropy', metrics=['accuracy'])
history_bin = model_bin.fit(x_train_flat, y_train_bin, epochs=5, batch_size=32, validation_split=0.2, verbose=0)

resultats_couts['binary_crossentropy'] = {
    'Précision': history_bin.history['val_accuracy'][-1],
    'Perte': history_bin.history['val_loss'][-1],
    'Description': "Binary Classification (ex. 5 ou non 5). Probabilité 0 ≤ y ≤ 1 [cite: 44]"
}

# ==========================================
# 2. categorical_crossentropy
# ==========================================
print("Entraînement avec categorical_crossentropy...")
model_cat = Sequential([
    Dense(128, activation='relu', input_shape=(784,)),
    # Sortie : 10 neurones avec softmax pour un vecteur des probabilités de chaque classe [cite: 44]
    Dense(10, activation='softmax')
])
model_cat.compile(optimizer='adam', loss='categorical_crossentropy', metrics=['accuracy'])
# Utilisation des labels encodés en one-hot (y_train_cat) [cite: 44]
history_cat = model_cat.fit(x_train_flat, y_train_cat, epochs=5, batch_size=32, validation_split=0.2, verbose=0)

resultats_couts['categorical_crossentropy'] = {
    'Précision': history_cat.history['val_accuracy'][-1],
    'Perte': history_cat.history['val_loss'][-1],
    'Description': "Multi-Class Classification. Vecteur des probabilités. À utiliser si les labels sont encodés en one-hot [cite: 44]"
}

# ==========================================
# 3. sparse_categorical_crossentropy
# ==========================================
print("Entraînement avec sparse_categorical_crossentropy...")
model_sparse = Sequential([
    Dense(128, activation='relu', input_shape=(784,)),
    # Sortie : 10 neurones avec softmax pour extraire les logits (raw scores) [cite: 44]
    Dense(10, activation='softmax')
])
model_sparse.compile(optimizer='adam', loss='sparse_categorical_crossentropy', metrics=['accuracy'])
# Utilisation des labels sous forme d'entiers (y_train), codés 0-9 [cite: 44]
history_sparse = model_sparse.fit(x_train_flat, y_train, epochs=5, batch_size=32, validation_split=0.2, verbose=0)

resultats_couts['sparse_categorical_crossentropy'] = {
    'Précision': history_sparse.history['val_accuracy'][-1],
    'Perte': history_sparse.history['val_loss'][-1],
    'Description': "Multi-Class (One Label). Logits (raw scores). À utiliser si les labels sont des entiers (codés 0-9) [cite: 44]"
}

# ==========================================
# GÉNÉRATION DU RÉSUMÉ DE TOUTES LES DONNÉES
# ==========================================
print("\n" + "="*100)
print(" RÉSUMÉ COMPARATIF DES FONCTIONS DE COÛT (APRÈS 5 EPOCHS)")
print("="*100)

for cout, metriques in resultats_couts.items():
    print(f"[{cout.upper()}]")
    print(f"  -> Précision (Val) : {metriques['Précision']:.4f}")
    print(f"  -> Perte (Val)     : {metriques['Perte']:.4f}")
    print(f"  -> Application     : {metriques['Description']}\n")

print("="*100)

Entraînement avec binary_crossentropy...
Entraînement avec categorical_crossentropy...
Entraînement avec sparse_categorical_crossentropy...

 RÉSUMÉ COMPARATIF DES FONCTIONS DE COÛT (APRÈS 5 EPOCHS)
[BINARY_CROSSENTROPY]
  -> Précision (Val) : 0.9912
  -> Perte (Val)     : 0.0324
  -> Application     : Binary Classification (ex. 5 ou non 5). Probabilité 0 ≤ y ≤ 1 [cite: 44]

[CATEGORICAL_CROSSENTROPY]
  -> Précision (Val) : 0.9720
  -> Perte (Val)     : 0.0924
  -> Application     : Multi-Class Classification. Vecteur des probabilités. À utiliser si les labels sont encodés en one-hot [cite: 44]

[SPARSE_CATEGORICAL_CROSSENTROPY]
  -> Précision (Val) : 0.9754
  -> Perte (Val)     : 0.0853
  -> Application     : Multi-Class (One Label). Logits (raw scores). À utiliser si les labels sont des entiers (codés 0-9) [cite: 44]

